images/virtualization-artifact/pkg/controller/vm/internal/pod.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2024 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+import (
+	"context"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/deckhouse/virtualization-controller/pkg/controller/powerstate"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/service"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vm/internal/state"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+)
+
+const namePodHandler = "PodHandler "
+
+func NewPodHandler(client client.Client) *PodHandler {
+	return &PodHandler{
+		client:     client,
+		protection: service.NewProtectionService(client, virtv2.FinalizerPodProtection),
+	}
+}
+
+type PodHandler struct {
+	client     client.Client
+	protection *service.ProtectionService
+}
+
+func (h *PodHandler) Handle(ctx context.Context, s state.VirtualMachineState) (reconcile.Result, error) {
+	if s.VirtualMachine().IsEmpty() {
+		return reconcile.Result{}, nil
+	}
+	current := s.VirtualMachine().Current()
+	pods, err := s.Pods(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if isDeletion(current) {
+		objs := make([]client.Object, len(pods.Items))
+		for i, p := range pods.Items {
+			objs[i] = p.DeepCopy()
+		}
+		return reconcile.Result{}, h.protection.RemoveProtection(ctx, objs...)
+	}
+	kvvmi, err := s.KVVMI(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	info := powerstate.ShutdownReason(kvvmi, pods)
+	if info.PodCompeted {
+		s.Shared(func(s *state.Shared) {
+			s.ShutdownInfo = info
+		})
```

Comment:
Do we need mutex guard here? PodHandler and SyncKvvmHandler are always executed one-by-one, never in parallel.

Reply:
So, you are right. Mb we don't need in mutex. this has been around for a long time and was added just in case for the future. 
Do you think we should remove this?

Reply:
Yes, I think you can get/set s.ShutdownInfo without locks.

Reply:
fixed